### PR TITLE
Interfaces lesson

### PR DESCRIPTION
**Pull request by zoefarrell**
Chat GPT did a hefty bit of the lifting here. I used it to generate an example (but did have to ask a few times to get an example I liked) and it also generated the additional questions students could look in to.

### CODE REVIEW

File: module4/lessons/Week4/Interfaces.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+* Part 2: Pair research to deepen your understanding of interfaces and how we have used it so far
+* Part 3: A pair presentation to your classmates about what you learned
+
+## Learning Goals
+
+* Understand the topic of interfaces
+* Identify how and why we have already used interfaces in MVC applications
+* Improve the skill of independent technical research and presenting on a technical topic
+
+## Part 0: Individual Reflection
```
Comment (memcmahon) [inline module4/lessons/Week4/Interfaces.md:23]: Love this reflection piece!!

File: module4/lessons/Week4/Interfaces.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+Finally, move on to digging into at least one more advanced topic. Here are some ideas, or feel free to dig into your own questions!
+
+1. What happens if a class doesn't implement all the methods of an interface in C#?
+1. What are some of the most commonly used interfaces in C# and .NET?
+1. Can a class implement multiple interfaces? If yes, how can you do that?
+1. How do you use an interface to achieve polymorphism in C#? (You will also need to google the term `polymorphism` for this one!)
+1. Can you define properties in an interface with different access levels (e.g., public get and private set)?
+1. Can you provide default implementations for interface methods in C#?
+1. Does an interface have a constructor? Why or why not?
+1. What is an abstract class? How is it different from an interface?
```
Comment (memcmahon) [inline module4/lessons/Week4/Interfaces.md:184]: I like that these research questions feel a bit more difficult than the inheritance ones!

File: module4/lessons/Week4/Interfaces.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+Each pair will have 7 minutes to present, and each partner should present for roughly equal time. Classmates will then have a couple of minutes to ask questions.
+
+1. Share at least one benefit of using interfaces and one potential disadvantage.
+1. Share one example of an interface from a past MVC project. Explain how and why an interface is used here.
+1. Explain one advanced part of interfaces you learned about today that was not covered in this morning's lesson (Could be one of the questions above or something else!). Use an example in your explanation.
+1. Share one resource you found useful while researching.
+
+
+## Final Checkpoint
```
Comment (memcmahon) [inline module4/lessons/Week4/Interfaces.md:198]: I think this is a good opportunity for some whole-group discussion.  The difference and use-cases for Inheritance and Interfaces can be very nuanced (at least that's how it seems to me).

**Review by memcmahon — APPROVED** [review]

File: module4/lessons/Week4/Interfaces.md (hunk 3 above)
Comment (zoefarrell) [reply module4/lessons/Week4/Interfaces.md:198]: Good call. I'll make an instructor note to run this think/pair/share style so it's more of a discussion than just a slack waterfall or something like that.

File: module4/lessons/Week4/Interfaces.md (hunk 1 above)
Comment (zoefarrell) [reply module4/lessons/Week4/Interfaces.md:23]: Thanks!

**Pull request merged**